### ISSUE TITLE
compile error when installing via pip3

### ISSUE THREAD

**Issue by atomfried** [roebel/pysndfile#4]
Hi,
I can't install via pip3 and I'll just use another tool because I don't have time to look into it but I thought I'd let you know... This is what I get:

> $ sudo pip3 install --upgrade pysndfile
Collecting pysndfile
  Using cached https://files.pythonhosted.org/packages/00/32/271564283bcde85c8b376290b65ed4eba3e10798a485cfd0ba80de94d993/pysndfile-1.3.3.tar.gz
Building wheels for collected packages: pysndfile
  Building wheel for pysndfile (setup.py) ... error
  ERROR: Complete output from command /usr/bin/python3 -u -c 'import setuptools, tokenize;__file__='"'"'/tmp/pip-install-x6c8t_ic/pysndfile/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-47e1zuc5 --python-tag cp37:
  ERROR: cythonize not available use pre_cythonized source
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build/lib.linux-x86_64-3.7
  creating build/lib.linux-x86_64-3.7/pysndfile
  copying pysndfile/Faiff.py -> build/lib.linux-x86_64-3.7/pysndfile
  copying pysndfile/__init__.py -> build/lib.linux-x86_64-3.7/pysndfile
  copying pysndfile/sndio.py -> build/lib.linux-x86_64-3.7/pysndfile
  running build_ext
  check for clang compiler ... no
  building '_pysndfile' extension
  creating build/temp.linux-x86_64-3.7
  x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I/usr/local/lib/python3.7/dist-packages/numpy/core/include -I/usr/include/python3.7m -c _pysndfile.cpp -o build/temp.linux-x86_64-3.7/_pysndfile.o
  In file included from /usr/local/lib/python3.7/dist-packages/numpy/core/include/numpy/ndarraytypes.h:1822,
                   from /usr/local/lib/python3.7/dist-packages/numpy/core/include/numpy/ndarrayobject.h:12,
                   from /usr/local/lib/python3.7/dist-packages/numpy/core/include/numpy/arrayobject.h:4,
                   from _pysndfile.cpp:449:
  /usr/local/lib/python3.7/dist-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: #warning "Using deprecated NumPy API, disable it with " "#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
   #warning "Using deprecated NumPy API, disable it with " \
    ^~~~~~~
  _pysndfile.cpp: In function ‘void __Pyx__ExceptionSave(PyThreadState*, PyObject**, PyObject**, PyObject**)’:
  _pysndfile.cpp:18554:21: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_type’; did you mean ‘curexc_type’?
       *type = tstate->exc_type;
                       ^~~~~~~~
                       curexc_type
  _pysndfile.cpp:18555:22: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_value’; did you mean ‘curexc_value’?
       *value = tstate->exc_value;
                        ^~~~~~~~~
                        curexc_value
  _pysndfile.cpp:18556:19: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_traceback’; did you mean ‘curexc_traceback’?
       *tb = tstate->exc_traceback;
                     ^~~~~~~~~~~~~
                     curexc_traceback
  _pysndfile.cpp: In function ‘void __Pyx__ExceptionReset(PyThreadState*, PyObject*, PyObject*, PyObject*)’:
  _pysndfile.cpp:18563:24: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_type’; did you mean ‘curexc_type’?
       tmp_type = tstate->exc_type;
                          ^~~~~~~~
                          curexc_type
  _pysndfile.cpp:18564:25: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_value’; did you mean ‘curexc_value’?
       tmp_value = tstate->exc_value;
                           ^~~~~~~~~
                           curexc_value
  _pysndfile.cpp:18565:22: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_traceback’; did you mean ‘curexc_traceback’?
       tmp_tb = tstate->exc_traceback;
                        ^~~~~~~~~~~~~
                        curexc_traceback
  _pysndfile.cpp:18566:13: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_type’; did you mean ‘curexc_type’?
       tstate->exc_type = type;
               ^~~~~~~~
               curexc_type
  _pysndfile.cpp:18567:13: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_value’; did you mean ‘curexc_value’?
       tstate->exc_value = value;
               ^~~~~~~~~
               curexc_value
  _pysndfile.cpp:18568:13: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_traceback’; did you mean ‘curexc_traceback’?
       tstate->exc_traceback = tb;
               ^~~~~~~~~~~~~
               curexc_traceback
  _pysndfile.cpp: In function ‘int __Pyx__GetException(PyThreadState*, PyObject**, PyObject**, PyObject**)’:
  _pysndfile.cpp:18623:24: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_type’; did you mean ‘curexc_type’?
       tmp_type = tstate->exc_type;
                          ^~~~~~~~
                          curexc_type
  _pysndfile.cpp:18624:25: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_value’; did you mean ‘curexc_value’?
       tmp_value = tstate->exc_value;
                           ^~~~~~~~~
                           curexc_value
  _pysndfile.cpp:18625:22: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_traceback’; did you mean ‘curexc_traceback’?
       tmp_tb = tstate->exc_traceback;
                        ^~~~~~~~~~~~~
                        curexc_traceback
  _pysndfile.cpp:18626:13: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_type’; did you mean ‘curexc_type’?
       tstate->exc_type = local_type;
               ^~~~~~~~
               curexc_type
  _pysndfile.cpp:18627:13: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_value’; did you mean ‘curexc_value’?
       tstate->exc_value = local_value;
               ^~~~~~~~~
               curexc_value
  _pysndfile.cpp:18628:13: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_traceback’; did you mean ‘curexc_traceback’?
       tstate->exc_traceback = local_tb;
               ^~~~~~~~~~~~~
               curexc_traceback
  _pysndfile.cpp: In function ‘PyObject* __Pyx_PyCFunction_FastCall(PyObject*, PyObject**, Py_ssize_t)’:
  _pysndfile.cpp:18707:69: error: too many arguments to function
       return (*((__Pyx_PyCFunctionFast)meth)) (self, args, nargs, NULL);
                                                                       ^
  error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
  ----------------------------------------
  ERROR: Failed building wheel for pysndfile
  Running setup.py clean for pysndfile
Failed to build pysndfile
Installing collected packages: pysndfile
  Running setup.py install for pysndfile ... error
    ERROR: Complete output from command /usr/bin/python3 -u -c 'import setuptools, tokenize;__file__='"'"'/tmp/pip-install-x6c8t_ic/pysndfile/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-devvjd2s/install-record.txt --single-version-externally-managed --compile:
    ERROR: cythonize not available use pre_cythonized source
    running install
    running build
    running build_py
    creating build
    creating build/lib.linux-x86_64-3.7
    creating build/lib.linux-x86_64-3.7/pysndfile
    copying pysndfile/Faiff.py -> build/lib.linux-x86_64-3.7/pysndfile
    copying pysndfile/__init__.py -> build/lib.linux-x86_64-3.7/pysndfile
    copying pysndfile/sndio.py -> build/lib.linux-x86_64-3.7/pysndfile
    running build_ext
    check for clang compiler ... no
    building '_pysndfile' extension
    creating build/temp.linux-x86_64-3.7
    x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I/usr/local/lib/python3.7/dist-packages/numpy/core/include -I/usr/include/python3.7m -c _pysndfile.cpp -o build/temp.linux-x86_64-3.7/_pysndfile.o
    In file included from /usr/local/lib/python3.7/dist-packages/numpy/core/include/numpy/ndarraytypes.h:1822,
                     from /usr/local/lib/python3.7/dist-packages/numpy/core/include/numpy/ndarrayobject.h:12,
                     from /usr/local/lib/python3.7/dist-packages/numpy/core/include/numpy/arrayobject.h:4,
                     from _pysndfile.cpp:449:
    /usr/local/lib/python3.7/dist-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: #warning "Using deprecated NumPy API, disable it with " "#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
     #warning "Using deprecated NumPy API, disable it with " \
      ^~~~~~~
    _pysndfile.cpp: In function ‘void __Pyx__ExceptionSave(PyThreadState*, PyObject**, PyObject**, PyObject**)’:
    _pysndfile.cpp:18554:21: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_type’; did you mean ‘curexc_type’?
         *type = tstate->exc_type;
                         ^~~~~~~~
                         curexc_type
    _pysndfile.cpp:18555:22: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_value’; did you mean ‘curexc_value’?
         *value = tstate->exc_value;
                          ^~~~~~~~~
                          curexc_value
    _pysndfile.cpp:18556:19: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_traceback’; did you mean ‘curexc_traceback’?
         *tb = tstate->exc_traceback;
                       ^~~~~~~~~~~~~
                       curexc_traceback
    _pysndfile.cpp: In function ‘void __Pyx__ExceptionReset(PyThreadState*, PyObject*, PyObject*, PyObject*)’:
    _pysndfile.cpp:18563:24: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_type’; did you mean ‘curexc_type’?
         tmp_type = tstate->exc_type;
                            ^~~~~~~~
                            curexc_type
    _pysndfile.cpp:18564:25: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_value’; did you mean ‘curexc_value’?
         tmp_value = tstate->exc_value;
                             ^~~~~~~~~
                             curexc_value
    _pysndfile.cpp:18565:22: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_traceback’; did you mean ‘curexc_traceback’?
         tmp_tb = tstate->exc_traceback;
                          ^~~~~~~~~~~~~
                          curexc_traceback
    _pysndfile.cpp:18566:13: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_type’; did you mean ‘curexc_type’?
         tstate->exc_type = type;
                 ^~~~~~~~
                 curexc_type
    _pysndfile.cpp:18567:13: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_value’; did you mean ‘curexc_value’?
         tstate->exc_value = value;
                 ^~~~~~~~~
                 curexc_value
    _pysndfile.cpp:18568:13: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_traceback’; did you mean ‘curexc_traceback’?
         tstate->exc_traceback = tb;
                 ^~~~~~~~~~~~~
                 curexc_traceback
    _pysndfile.cpp: In function ‘int __Pyx__GetException(PyThreadState*, PyObject**, PyObject**, PyObject**)’:
    _pysndfile.cpp:18623:24: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_type’; did you mean ‘curexc_type’?
         tmp_type = tstate->exc_type;
                            ^~~~~~~~
                            curexc_type
    _pysndfile.cpp:18624:25: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_value’; did you mean ‘curexc_value’?
         tmp_value = tstate->exc_value;
                             ^~~~~~~~~
                             curexc_value
    _pysndfile.cpp:18625:22: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_traceback’; did you mean ‘curexc_traceback’?
         tmp_tb = tstate->exc_traceback;
                          ^~~~~~~~~~~~~
                          curexc_traceback
    _pysndfile.cpp:18626:13: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_type’; did you mean ‘curexc_type’?
         tstate->exc_type = local_type;
                 ^~~~~~~~
                 curexc_type
    _pysndfile.cpp:18627:13: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_value’; did you mean ‘curexc_value’?
         tstate->exc_value = local_value;
                 ^~~~~~~~~
                 curexc_value
    _pysndfile.cpp:18628:13: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘exc_traceback’; did you mean ‘curexc_traceback’?
         tstate->exc_traceback = local_tb;
                 ^~~~~~~~~~~~~
                 curexc_traceback
    _pysndfile.cpp: In function ‘PyObject* __Pyx_PyCFunction_FastCall(PyObject*, PyObject**, Py_ssize_t)’:
    _pysndfile.cpp:18707:69: error: too many arguments to function
         return (*((__Pyx_PyCFunctionFast)meth)) (self, args, nargs, NULL);
                                                                         ^
    error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
    ----------------------------------------
ERROR: Command "/usr/bin/python3 -u -c 'import setuptools, tokenize;__file__='"'"'/tmp/pip-install-x6c8t_ic/pysndfile/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-devvjd2s/install-record.txt --single-version-externally-managed --compile" failed with error code 1 in /tmp/pip-install-x6c8t_ic/pysndfile/

**Comment by roebel** [comment]
Thanks, I prepared this with python3.6 and you have python 3.7, You need to install cython for your python installation for this to work. I will look how to add this dependency into the installer.

**Comment by sveinse** [comment]
I had some similar problems when using pysndfile from pypi. Currently the package assumes and requires numpy and cython, which is not perfect. E.g. running

```
$ pip download --no-deps --no-binary=:all pysndfile
Collecting pysndfile
  Using cached https://files.pythonhosted.org/packages/00/32/271564283bcde85c8b376290b65ed4eba3e10798a485cfd0ba80de94d993/pysndfile-1.3.3.tar.gz
  Saved c:\svein\prosjekt\elns\pysndfile-1.3.3.tar.gz
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "C:\Users\svein\AppData\Local\Temp\pip-download-9vz9dlkp\pysndfile\setup.py", line 11, in <module>
        import numpy as np
    ModuleNotFoundError: No module named 'numpy'
```

fails for simple download. The expected behavior is that pip should be able to download the source repo. This is caused by the direct import of numpy inside the setup.py script. Unfortunately, I haven't any knowledge on how it should be implemented. I pushed the problem ahead of myself by simply preinstalling numpy and cython in the venv before dealing with pysndfile.

**Comment by roebel** [comment]
The problem is that when I started the project in 2014 build requirements could not be handled by setup.py. Moreover there is always the dependency of libsndfile that cannot be handled anyway because it is outside the pypi world. 

While libsndfile is still not handled there exist means to handle build dependencies (here cython and numpy) following suggestions in [PEP 518](https://www.python.org/dev/peps/pep-0518/). I made a few trials and at least for the latest version of  _pip 19.1_ this seems to work quite nicely. You find this implemented in pysndfile 1.3.4. It includes as well the changes from issue https://github.com/roebel/pysndfile/issues/3 related to the support for windows.  If I don't get any feedback for problems here I will close this issue.
